src/wp-includes/block-bindings/class-wp-block-bindings.php

```diff
@@ -0,0 +1,281 @@
+<?php
+/**
+ * Block Bindings API: WP_Block_Bindings class.
+ *
+ * Support for overriding content in blocks by connecting them to different sources.
+ *
+ * @package WordPress
+ * @subpackage Block Bindings
+ */
+
+/**
+ * Core class used to define supported blocks, register sources, and populate HTML with content from those sources.
+ *
+ *  @since 6.5.0
+ */
+class WP_Block_Bindings_Registry {
+
+	/**
+	 * Holds the registered block bindings sources, keyed by source identifier.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @var array
+	 */
+	private $sources = array();
+
+	/**
+	 * Function to register a new block binding source.
+	 *
+	 * Sources are used to override block's original attributes with a value
+	 * coming from the source. Once a source is registered, it can be used by a
+	 * block by setting its `metadata.bindings` attribute to a value that refers
+	 * to the source.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string   $source_name   The name of the source.
+	 * @param array    $source_args   The array of arguments that are used to register a source. The array has two elements:
+	 *                                1. string   $label        The label of the source.
+	 *                                2. callback $apply        A callback
+	 *                                executed when the source is processed during
+	 *                                block rendering. The callback should have the
+	 *                                following signature:
+	 *
+	 *                                  `function (object $source_attrs, object $block_instance, string $attribute_name): string`
+	 *                                          - @param object $source_attrs: Object containing source ID used to look up the override value, i.e. {"value": "{ID}"}.
+	 *                                          - @param object $block_instance: The block instance.
+	 *                                          - @param string $attribute_name: The name of an attribute used to retrieve an override value from the block context.
+	 *                                 The callback should return a string that will be used to override the block's original value.
+	 *
+	 * @return void
+	 */
+	public function register_block_bindings_source( $source_name, array $source_args ) {
```

Comment:
It would be great to add validation to avoid:
- registering the same source twice
- namaspacing would be useful like in all other registries, in effect we would have: `core/pattern-overrides` or `core/post-meta`, but plugins would use `acf/post-meta`, etc.

Reply:
Agreed 👍

Reply:
Ok, I added the validation in https://github.com/WordPress/wordpress-develop/pull/5888/commits/2900bf53330d355f476bce07f04f36316618b4ee

How does this look? I'm not sure if we should fire an error or warning if a source has already been registered.
Also, I saw that the `WP_Block_Styles_Registry` [returns a boolean](https://github.com/WordPress/wordpress-develop/blob/e42e65def6ddad2c9a100e303c06b39aaff733d0/src/wp-includes/class-wp-block-styles-registry.php#L98) in the registration function, so I thought that approach might be good to emulate here.

Reply:
That’s good for start 👍